camel/data_collector/base.py

```diff
@@ -0,0 +1,169 @@
+# ========= Copyright 2023-2024 @ CAMEL-AI.org. All Rights Reserved. =========
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ========= Copyright 2023-2024 @ CAMEL-AI.org. All Rights Reserved. =========
+
+import uuid
+from abc import ABC, abstractmethod
+from typing import Any, Dict, List, Optional, Tuple, Union
+from uuid import UUID
+
+from typing_extensions import Self
+
+from camel.agents import ChatAgent
+from camel.messages.base import BaseMessage
+from camel.types.enums import OpenAIBackendRole
+
+
+class CollectorData:
+    def __init__(
+        self,
+        id: UUID,
+        name: str,
+        role: OpenAIBackendRole,
+        message: BaseMessage,
+    ) -> None:
+        self.id = id
+        self.name = name
+        self.role = role
+        self.message = message
+
+
+class BaseDataCollector(ABC):
+    r"""Base class for data collectors."""
+
+    def __init__(self) -> None:
+        self.history: List[CollectorData] = []
+        self._recording = False
+        self.agents: List[Tuple[str, ChatAgent]] = []
+        self.data: List[Dict[str, Any]] = []
+
+    def step(
+        self,
+        message: BaseMessage,
+        role: OpenAIBackendRole,
+        role_name: Optional[str] = None,
+    ) -> Self:
+        r"""Record a message.
+
+        Args:
+            message (Union[BaseMessage, ChatAgentResponse]):
+                The message to record.
+            role (OpenAIBackendRole): The role of the message.
```

Comment:
`role` can also be optional, as we could get `RoleType` from `message.role_type`, then classify it into `OpenAIBackendRole `